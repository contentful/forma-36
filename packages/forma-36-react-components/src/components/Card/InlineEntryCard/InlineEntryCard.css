@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.InlineEntryCard {
  max-width: 100%;
  display: inline-flex;
  align-items: center;
  line-height: 1.5;
  position: relative;
  padding: 0px calc(1rem * (4 / var(--font-base-default))) 0px
    calc(1rem * (12 / var(--font-base-default)));
  color: inherit;
  font-size: inherit;
  font-weight: inherit;
  font-family: var(--font-stack-primary);
  transition: width var(--transition-duration-default)
    var(--transition-easing-default);
}

.InlineEntryCard__skeleton-wrapper {
  position: absolute;
  width: 100%;
  height: 100%;
  background: var(--color-white);
  z-index: var(--z-index-default);
  left: 0px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  padding: 0 calc(1rem * (3 / var(--font-base-default)));
}

.InlineEntryCard__text-wrapper {
  max-height: 1.5em;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.InlineEntryCard__status-indicator {
  width: calc(1rem * (4 / var(--font-base-default)));
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
}

.InlineEntryCard__status-indicator--published {
  outline: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-green-dark);
  background: var(--color-green-mid);
}

.InlineEntryCard__status-indicator--draft {
  outline: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-orange-dark);
  background: var(--color-orange-mid);
}

.InlineEntryCard__status-indicator--archived {
  outline: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-red-dark);
  background: var(--color-red-mid);
}

.InlineEntryCard__status-indicator--changed {
  outline: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-blue-dark);
  background: var(--color-blue-mid);
}

.InlineEntryCard__status-indicator {
  height: 100%;
  width: calc(1rem * (4 / var(--font-base-default)));
}

.InlineEntryCard__actions {
  display: inline-flex;
  position: static;
  margin-left: calc(1rem * (6 / var(--font-base-default)));
}

.InlineEntryCard__spinner--enter {
  opacity: 0;
}

.InlineEntryCard__spinner--enter-active {
  opacity: 1;
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
}

.InlineEntryCard__spinner--exit {
  opacity: 1;
}

.InlineEntryCard__spinner--exit-active {
  opacity: 0;
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
}
