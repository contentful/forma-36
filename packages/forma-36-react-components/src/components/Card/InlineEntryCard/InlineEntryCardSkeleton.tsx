import React from 'react';

import { SkeletonBodyText, SkeletonContainer } from '../../Skeleton';

export const InlineEntryCardSkeleton = () => (
  <SkeletonContainer
    width="100%"
    clipId="f36-inline-entry-card-skeleton"
    svgHeight={16}
  >
    <SkeletonBodyText numberOfLines={1} />
  </SkeletonContainer>
);
