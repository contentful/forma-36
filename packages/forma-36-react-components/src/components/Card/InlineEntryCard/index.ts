export * from './InlineEntryCard';
export { default } from './InlineEntryCard';
