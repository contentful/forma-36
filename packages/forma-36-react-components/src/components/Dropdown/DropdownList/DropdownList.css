@import 'settings/colors';
@import 'settings/typography';

.DropdownList {
  padding: 0;
  margin: 0;
  padding: calc(1rem * (14 / var(--font-base-default))) 0;
}

/* For multiple DropdownLists */
.DropdownList + .DropdownList {
  padding: calc(1rem * (7 / var(--font-base-default))) 0;
}

/* First DropdownList if multiple */
.DropdownList:first-child:not(:only-child) {
  padding-bottom: calc(1rem * (7 / var(--font-base-default)));
}

/* Last DropdownList if multiple */
.DropdownList:last-child:not(:only-child) {
  padding-bottom: calc(1rem * (14 / var(--font-base-default)));
}

.DropdownList--border-top {
  border-top: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-element-mid);
}

.DropdownList--border-bottom {
  border-bottom: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-element-mid);
}
