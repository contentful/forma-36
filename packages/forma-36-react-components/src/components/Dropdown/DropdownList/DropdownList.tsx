import React from 'react';
import cn from 'classnames';
import type * as CSS from 'csstype';

import cssStyles from './DropdownList.css';

export interface DropdownListProps {
  children: React.ReactNode;
  listRef?: React.RefObject<HTMLUListElement>;
  className?: string;
  testId?: string;
  border?: 'top' | 'bottom';
  maxHeight?: number | CSS.Property.MaxHeight;
  styles?: object;
}

export const DropdownList = ({
  className,
  border,
  maxHeight,
  testId,
  children,
  listRef,
  styles,
  ...otherProps
}: DropdownListProps) => {
  const classNames = cn(cssStyles['DropdownList'], className, {
    [cssStyles[`DropdownList--border-${border}`]]: border,
  });

  return (
    <ul
      ref={listRef}
      data-test-id={testId}
      role="menu"
      style={{
        maxHeight: maxHeight || 'auto',
        overflowY: maxHeight ? 'auto' : 'visible',
        ...styles,
      }}
      className={classNames}
      {...otherProps}
    >
      {children}
    </ul>
  );
};

DropdownList.defaultProps = {
  testId: 'cf-ui-dropdown-list',
};

export default DropdownList;
