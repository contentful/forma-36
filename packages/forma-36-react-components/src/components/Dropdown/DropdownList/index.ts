export * from './DropdownList';
export { default } from './DropdownList';
