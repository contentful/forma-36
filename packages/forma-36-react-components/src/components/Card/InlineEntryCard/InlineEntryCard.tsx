import React from 'react';
import cn from 'classnames';
import { CSSTransition } from 'react-transition-group';
import type { EntityStatus, PickUnion } from '@contentful/f36-core';

import { Card } from '../Card';
import { CardActions } from '../CardActions/CardActions';
import { InlineEntryCardSkeleton } from './InlineEntryCardSkeleton';
import styles from './InlineEntryCard.css';

export type InlineEntryCardStatus = PickUnion<
  EntityStatus,
  'archived' | 'changed' | 'draft' | 'published'
>;

export interface InlineEntryCardProps {
  /**
   * Gives the component a selected state
   */
  isSelected?: boolean;
  /**
   * The DropdownList elements used to render an actions dropdown for the component
   */
  dropdownListElements?: React.ReactElement;
  /**
   * Loading state for the component - when true will display loading feedback to the user
   */
  isLoading?: boolean;
  /**
   * The publish status of the entry
   */
  status?: InlineEntryCardStatus;

  /**
   * Class names to be appended to the className prop of the component
   */
  className?: string;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id)
   */
  testId?: string;
  href?: string;
  /**
   * Child nodes to be rendered in the component
   */
  children: React.ReactNode;
}

export function InlineEntryCard({
  className,
  dropdownListElements,
  isSelected,
  children,
  testId = 'cf-ui-inline-entry-card',
  isLoading,
  status,
  href,
  ...otherProps
}: InlineEntryCardProps): React.ReactElement {
  const classNames = cn(styles.InlineEntryCard, className);

  const statusIndicatorClassNames = cn(
    styles['InlineEntryCard__status-indicator'],
    {
      [styles[`InlineEntryCard__status-indicator--${status}`]]:
        status && !isLoading,
    },
  );

  return (
    <Card
      selected={isSelected}
      className={classNames}
      href={href}
      {...otherProps}
      data-test-id={testId}
    >
      <CSSTransition
        timeout={100}
        in={isLoading}
        classNames={{
          enter: styles['InlineEntryCard__spinner--enter'],
          enterActive: styles['InlineEntryCard__spinner--enter-active'],
          exit: styles['InlineEntryCard__spinner--exit'],
          exitActive: styles['InlineEntryCard__spinner--exit-active'],
        }}
        mountOnEnter
        unmountOnExit
      >
        <div className={styles['InlineEntryCard__skeleton-wrapper']}>
          <InlineEntryCardSkeleton />
        </div>
      </CSSTransition>
      <div className={statusIndicatorClassNames} />
      <span className={styles['InlineEntryCard__text-wrapper']}>
        {isLoading ? 'Loading' : children}
      </span>
      {dropdownListElements && (
        <CardActions
          className={styles['InlineEntryCard__actions']}
          iconButtonProps={{
            onClick: (e) => e.stopPropagation,
          }}
        >
          {dropdownListElements}
        </CardActions>
      )}
    </Card>
  );
}
