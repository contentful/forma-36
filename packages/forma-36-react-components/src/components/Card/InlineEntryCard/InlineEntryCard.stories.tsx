import React from 'react';

import { InlineEntryCard, InlineEntryCardPropTypes } from './InlineEntryCard';
import { DropdownList, DropdownListItem } from '../../Dropdown';

export default {
  title: 'Components/Card/InlineEntryCard',
  component: InlineEntryCard,
  parameters: {
    propTypes: [InlineEntryCard['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

const dropdownElements = (
  <React.Fragment>
    <DropdownList>
      <DropdownListItem isTitle>Actions</DropdownListItem>
      <DropdownListItem href="#">Edit (with href)</DropdownListItem>
      <DropdownListItem onClick={() => {}}>Download</DropdownListItem>
      <DropdownListItem onClick={() => {}}>Remove</DropdownListItem>
    </DropdownList>
    <DropdownList>
      <DropdownListItem isTitle>Actions</DropdownListItem>
      <DropdownListItem onClick={() => {}}>Edit</DropdownListItem>
      <DropdownListItem onClick={() => {}}>Download</DropdownListItem>
      <DropdownListItem onClick={() => {}}>Remove</DropdownListItem>
    </DropdownList>
  </React.Fragment>
);

interface Args extends InlineEntryCardPropTypes {
  entryText?: string;
}

export const Basic = ({ entryText, ...args }: Args) => (
  <div
    style={{
      lineHeight: 1.5,
      display: 'inline-block',
      maxWidth: 600,
      fontSize: 16,
    }}
  >
    Fusce a odio pharetra, porta justo vel, maximus ex. In pellentesque a orci
    et pretium. Praesent libero lorem, gravida eu pulvinar id, eleifend a
    sapien. &nbsp;
    <InlineEntryCard {...args}>{entryText}</InlineEntryCard>
    &nbsp; Nulla a ultrices nulla, vel blandit sapien. Etiam eget massa dui.
    Fusce id nisl quam. Integer nec mi arcu. Nullam lacinia est lectus, a
    euismod purus eleifend id. Fusce a odio pharetra, porta justo vel, maximus
    ex. In pellentesque a orci et pretium. Praesent libero lorem, gravida eu
    pulvinar id, eleifend a sapien. &nbsp;
    <InlineEntryCard {...args}>{entryText}</InlineEntryCard>
    &nbsp; Nulla a ultrices nulla, vel blandit sapien. Etiam eget massa dui.
    Fusce id nisl quam. Integer nec mi arcu. Nullam lacinia est lectus, a
    euismod purus eleifend id.
  </div>
);

Basic.args = {
  status: 'published',
  dropdownListElements: dropdownElements,
  entryText: 'Title of inline entry',
};
