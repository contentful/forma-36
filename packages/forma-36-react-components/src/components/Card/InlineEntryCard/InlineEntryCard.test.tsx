import React from 'react';
import { shallow, mount } from 'enzyme';
import axe from '../../../utils/axeHelper';
import InlineEntryCard from './InlineEntryCard';
import DropdownListItem from '../../Dropdown/DropdownListItem';
import DropdownList from '../../Dropdown/DropdownList';

it('renders the component', () => {
  const output = shallow(
    <InlineEntryCard>Title of inline entry</InlineEntryCard>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <InlineEntryCard className="my-extra-class">
      Title of inline entry
    </InlineEntryCard>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with published status', () => {
  const output = shallow(
    <InlineEntryCard className="my-extra-class" status="published">
      Title of inline entry
    </InlineEntryCard>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with a dropdown', () => {
  const output = shallow(
    <InlineEntryCard
      className="my-extra-class"
      dropdownListElements={
        <DropdownList>
          <DropdownListItem key="edit" onClick={() => {}}>
            Edit
          </DropdownListItem>
          <DropdownListItem key="remove" onClick={() => {}}>
            Remove
          </DropdownListItem>
        </DropdownList>
      }
    >
      Title of inline entry
    </InlineEntryCard>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <InlineEntryCard>InlineEntryCard</InlineEntryCard>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
