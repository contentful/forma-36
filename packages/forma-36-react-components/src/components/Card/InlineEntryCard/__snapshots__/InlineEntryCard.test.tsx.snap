// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="Card InlineEntryCard Card--padding-default"
  data-test-id="cf-ui-inline-entry-card"
>
  <div
    class="InlineEntryCard__status-indicator"
  />
  <span
    class="InlineEntryCard__text-wrapper"
  >
    Title of inline entry
  </span>
</div>
`;

exports[`renders the component with a dropdown 1`] = `
.emotion-1 {
  display: inherit;
  outline: 0;
}

.emotion-0 {
  display: inline-block;
  fill: #2D64B3;
  height: 18px;
  width: 18px;
  display: inline-block;
  fill: #2D64B3;
  height: 18px;
  width: 18px;
}

<div
  class="Card InlineEntryCard my-extra-class Card--padding-default"
  data-test-id="cf-ui-inline-entry-card"
>
  <div
    class="InlineEntryCard__status-indicator"
  />
  <span
    class="InlineEntryCard__text-wrapper"
  >
    Title of inline entry
  </span>
  <div
    class="Dropdown InlineEntryCard__actions"
    data-test-id="cf-ui-card-actions"
  >
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      class="IconButton IconButton--secondary"
      data-test-id="cf-ui-icon-button"
      type="button"
    >
      <span
        class="emotion-1 IconButton__inner"
        tabindex="-1"
      >
        <svg
          class="IconButton__icon emotion-0"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
        <span
          class="IconButton__label"
        >
          Actions
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="Card InlineEntryCard my-extra-class Card--padding-default"
  data-test-id="cf-ui-inline-entry-card"
>
  <div
    class="InlineEntryCard__status-indicator"
  />
  <span
    class="InlineEntryCard__text-wrapper"
  >
    Title of inline entry
  </span>
</div>
`;

exports[`renders the component with published status 1`] = `
<div
  class="Card InlineEntryCard my-extra-class Card--padding-default"
  data-test-id="cf-ui-inline-entry-card"
>
  <div
    class="InlineEntryCard__status-indicator InlineEntryCard__status-indicator--published"
  />
  <span
    class="InlineEntryCard__text-wrapper"
  >
    Title of inline entry
  </span>
</div>
`;
